domain/port/service/service_test.go

```diff
@@ -0,0 +1,263 @@
+// Copyright 2024 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package service
+
+import (
+	"context"
+
+	jc "github.com/juju/testing/checkers"
+	"go.uber.org/mock/gomock"
+	gc "gopkg.in/check.v1"
+
+	"github.com/juju/juju/core/errors"
+	"github.com/juju/juju/core/network"
+	domain "github.com/juju/juju/domain"
+)
+
+type serviceSuite struct {
+	st *MockState
+}
+
+var _ = gc.Suite(&serviceSuite{})
+
+const unitUUID = "unit-uuid"
+
+func (s *serviceSuite) setupMocks(c *gc.C) *gomock.Controller {
+	ctrl := gomock.NewController(c)
+
+	s.st = NewMockState(ctrl)
+	s.st.EXPECT().RunAtomic(gomock.Any(), gomock.Any()).DoAndReturn(func(ctx context.Context, fn func(ctx domain.AtomicContext) error) error {
+		return fn(nil)
+	}).AnyTimes()
+
+	return ctrl
+}
+
+func (s *serviceSuite) TestGetOpenedPorts(c *gc.C) {
+	defer s.setupMocks(c).Finish()
+
+	grp := network.GroupedPortRanges{
+		"ep1": {
+			network.MustParsePortRange("80/tcp"),
+			network.MustParsePortRange("443/tcp"),
+		},
+		"ep2": {
+			network.MustParsePortRange("8000-9000/udp"),
+		},
+	}
+
+	s.st.EXPECT().GetOpenedPorts(gomock.Any(), unitUUID).Return(grp, nil)
+
+	srv := NewService(s.st)
+	res, err := srv.GetOpenedPorts(context.Background(), unitUUID)
+	c.Assert(err, gc.IsNil)
+	c.Assert(res, gc.DeepEquals, grp)
+}
+
+func (s *serviceSuite) TestUpdateUnitPorts(c *gc.C) {
+	defer s.setupMocks(c).Finish()
+
+	openPorts := network.GroupedPortRanges{
+		"ep1": {
+			network.MustParsePortRange("80/tcp"),
+			network.MustParsePortRange("443/tcp"),
+		},
+		"ep2": {
+			network.MustParsePortRange("8000-9000/udp"),
+		},
+	}
+
+	closePorts := network.GroupedPortRanges{
+		"ep1": {
+			network.MustParsePortRange("22/tcp"),
+		},
+	}
+
+	s.st.EXPECT().GetOpenedEndpointPorts(gomock.Any(), unitUUID, WildcardEndpoint).Return([]network.PortRange{}, nil)
+	s.st.EXPECT().UpdateUnitPorts(gomock.Any(), unitUUID, openPorts, closePorts).Return(nil)
```

Comment:
Can you change these to a matcher returned from `AssignableToTypeOf`?

You don't need to create it anew every time, you could create a variable in package_test.

Reply:
This will require more work around testing with RunAtomic. I will submit a follow up PR